crates/mongodb-connector/src/query_response.rs

```diff
@@ -0,0 +1,815 @@
+        responses
+            .row_sets
+            .into_iter()
+            .map(|docs| {
+                serialize_row_set(
+                    query_context,
+                    query_request,
+                    &[],
+                    collection_name,
+                    &query_request.query,
+                    docs,
+                )
+            })
+            .try_collect()
+    } else {
+        Ok(vec![serialize_row_set(
+            query_context,
+            query_request,
+            &[],
+            collection_name,
+            &query_request.query,
+            response_documents,
+        )?])
+    }?;
+    let response = QueryResponse(row_sets);
+    tracing::debug!(query_response = %serde_json::to_string(&response).unwrap());
+    Ok(response)
+}
+
+fn serialize_row_set(
+    query_context: &QueryContext<'_>,
+    query_request: &QueryRequest,
+    path: &[&str],
+    collection_name: &str,
+    query: &Query,
+    docs: Vec<bson::Document>,
+) -> Result<RowSet> {
+    if query
+        .aggregates
+        .as_ref()
+        .unwrap_or(&IndexMap::new())
+        .is_empty()
+    {
+        // When there are no aggregates we expect a list of rows
+        let rows = query
+            .fields
+            .as_ref()
+            .map(|fields| {
+                serialize_rows(
+                    query_context,
+                    query_request,
+                    path,
+                    collection_name,
+                    fields,
+                    docs,
+                )
+            })
+            .transpose()?;
+        Ok(RowSet {
+            aggregates: None,
+            rows,
+        })
+    } else {
+        // When there are aggregates we expect a single document with `rows` and `aggregates`
+        // fields
+        let row_set: BsonRowSet = parse_single_document(docs)?;
+
+        let aggregates = query
+            .aggregates
+            .as_ref()
+            .map(|aggregates| serialize_aggregates(aggregates, row_set.aggregates))
+            .transpose()?;
+
+        let rows = query
+            .fields
+            .as_ref()
+            .map(|fields| {
+                serialize_rows(
+                    query_context,
+                    query_request,
+                    path,
+                    collection_name,
+                    fields,
+                    row_set.rows,
+                )
+            })
+            .transpose()?;
+
+        Ok(RowSet { aggregates, rows })
+    }
+}
+
+fn serialize_aggregates(
+    query_aggregates: &IndexMap<String, Aggregate>,
+    mut aggregate_values: BTreeMap<String, Bson>,
+) -> Result<IndexMap<String, serde_json::Value>> {
+    query_aggregates
+        .iter()
+        .map(
+            |(key, _aggregate_definition)| match aggregate_values.remove_entry(key) {
+                Some((owned_key, value)) => Ok((
+                    owned_key,
+                    // TODO: bson_to_json
```

Comment:
Also later TODO or for this PR?

Reply:
Filed https://hasurahq.atlassian.net/browse/MDB-130 to follow up on this